packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +468,521 @@ class CommonFinders {
+  FinderBase<SemanticsNode> byAction(SemanticsAction action, {FlutterView? view}) {
+    return byPredicate(
+      (SemanticsNode node) => node.getSemanticsData().hasAction(action),
+      requirements: 'with action "$action"',
+      view: view,
+    );
+  }
+
+  /// Finds any [SemanticsNode]s that has at least one of the given
+  /// [SemanticsAction]s.
+  ///
+  /// {@macro flutter_test.finders.CommonSemanticsFinders.viewParameter}
+  FinderBase<SemanticsNode> byAnyAction(List<SemanticsAction> actions, {FlutterView? view}) {
+    final int actionsInt = actions.fold(0, (int value, SemanticsAction action) => value | action.index);
+    return byPredicate(
+      (SemanticsNode node) => node.getSemanticsData().actions & actionsInt != 0,
+      requirements: 'with any of the following actions: $actions',
+      view: view,
+    );
+  }
+
+  /// Finds any [SemanticsNode]s that has the given [SemanticsFlag].
+  ///
+  /// {@macro flutter_test.finders.CommonSemanticsFinders.viewParameter}
+  FinderBase<SemanticsNode> byFlag(SemanticsFlag flag, {FlutterView? view}) {
+    return byPredicate(
+      (SemanticsNode node) => node.hasFlag(flag),
+      requirements: 'with flag "$flag"',
+      view: view,
+    );
+  }
+
+  /// Finds any [SemanticsNode]s that has at least one of the given
+  /// [SemanticsFlag]s.
+  ///
+  /// {@macro flutter_test.finders.CommonSemanticsFinders.viewParameter}
+  FinderBase<SemanticsNode> byAnyFlag(List<SemanticsFlag> flags, {FlutterView? view}) {
+    final int flagsInt = flags.fold(0, (int value, SemanticsFlag flag) => value | flag.index);
+    return byPredicate(
+      (SemanticsNode node) => node.getSemanticsData().flags & flagsInt != 0,
+      requirements: 'with any of the following flags: $flags',
+      view: view,
+    );
+  }
+
+  bool _matchesPattern(String target, Pattern pattern) {
+    if (pattern is RegExp) {
+      return pattern.hasMatch(target);
+    } else {
+      return pattern == target;
+    }
+  }
+
+  SemanticsNode _rootFromView(FlutterView? view) {
+    view ??= TestWidgetsFlutterBinding.instance.platformDispatcher.implicitView;
+
+    // TODO(pdblasi-google): Update this to get a view-specific semantics root when available.
+    return TestWidgetsFlutterBinding.instance.pipelineOwner.semanticsOwner!.rootSemanticsNode!;
+  }
+}
+
+/// Encapsulates the logic for searching a list of candidates and filtering the
+/// candidates to only those that meet the requirements defined by the finder.
+///
+/// Implementations will need to implement [allCandidates] to define the total
+/// possible search space and [findInCandidates] to define the requirements of
+/// the finder.
+///
+/// This library contains [ElementFinder], [SemanticsFinder], and mixins of each
+/// for searching Flutter's element and semantics trees respectively. If the
+/// search can be represented as a predicate, then use [MatchFinderMixin] along
+/// with the appropriate base class.
+abstract class FinderBase<TCandidate> {
```

Comment:
IMHO `AbstractFinder` or `FinderDescription` (or just `Finder` - but that'd be too breaking I suppose) is a bit better name, but no strong opinions here.

Reply:
I'd've _loved_ to reuse `Finder`, but you're right that it's too breaking.

Reply:
When (the old, deprecated) `Finder` is removed from the codebase, can a `typedef Finder = FinderBase` be added?

Reply:
We generally try to avoid four step deprecations (see the [Tree Hygiene page on the topic](https://github.com/flutter/flutter/wiki/Tree-hygiene#3-prepare-your-change)) to avoid breaking upgrades that jump across the versions where the symbol existed but were different.

Reply:
Ah right. Makes sense!